src/errors.rs

```diff
@@ -16,48 +16,77 @@ pub enum Error {
     PermanentFailure { message: String },
 }
 
-pub fn invalid_input<T: ToString>(e: T) -> Error {
+pub fn invalid_input<E: ToString>(e: E) -> Error {
     Error::InvalidInput {
         message: e.to_string(),
     }
 }
 
-pub fn invalid_input_with<T: ToString>(message: String) -> Box<dyn FnOnce(T) -> Error> {
-    Box::new(move |e: T| Error::InvalidInput {
-        message: format!("{}: {}", message, e.to_string()),
+pub fn invalid_input_with<M: ToString + 'static, E: ToString>(
+    message: M,
+) -> Box<dyn FnOnce(E) -> Error> {
+    Box::new(move |e: E| Error::InvalidInput {
+        message: format!("{}: {}", message.to_string(), e.to_string()),
     })
 }
 
-pub fn runtime_error<T: ToString>(e: T) -> Error {
+pub fn runtime_error<E: ToString>(e: E) -> Error {
     Error::RuntimeError {
         message: e.to_string(),
     }
 }
 
-pub fn runtime_error_with<T: ToString>(message: String) -> Box<dyn FnOnce(T) -> Error> {
-    Box::new(move |e: T| Error::RuntimeError {
-        message: format!("{}: {}", message, e.to_string()),
+pub fn runtime_error_with<M: ToString + 'static, E: ToString>(
+    message: M,
+) -> Box<dyn FnOnce(E) -> Error> {
+    Box::new(move |e: E| Error::RuntimeError {
+        message: format!("{}: {}", message.to_string(), e.to_string()),
     })
 }
 
-pub fn permanent_failure<T: ToString>(e: T) -> Error {
+pub fn permanent_failure<E: ToString>(e: E) -> Error {
     Error::PermanentFailure {
         message: e.to_string(),
     }
 }
 
-pub fn permanent_failure_with<T: ToString>(message: String) -> Box<dyn FnOnce(T) -> Error> {
-    Box::new(move |e: T| Error::PermanentFailure {
-        message: format!("{}: {}", message, e.to_string()),
+pub fn permanent_failure_with<M: ToString + 'static, E: ToString>(
+    message: M,
+) -> Box<dyn FnOnce(E) -> Error> {
+    Box::new(move |e: E| Error::PermanentFailure {
+        message: format!("{}: {}", message.to_string(), e.to_string()),
     })
 }
 
-pub fn lift_invalid_input(e: Error) -> Error {
-    match e {
-        Error::InvalidInput { message } => Error::PermanentFailure {
-            message: "InvalidInput: ".to_string() + &message,
-        },
-        another_error => another_error,
+pub type Result<T> = std::result::Result<T, Error>;
```

Comment:
I would rename this to `LipaResult` and the trait to another name. I think it can be confusing that the `Result` we use is not the std one.